pandas_gbq/gbq.py

```diff
@@ -956,61 +803,148 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
+            else:
+                bytes_billed = int(query_job._properties
+                                   .get("totalBytesProcessed", 0))
+                bytes_processed = int(query_job._properties
+                                      .get("totalBytesBilled", 0))
+                print("Total bytes billed (processed): %s (%s)" %
+                      (sizeof_fmt(bytes_billed), sizeof_fmt(bytes_processed)))
+            print("\nRetrieving results...")
+        return query_job, None
+
+    def async_query():
+        query_job = client.run_async_query(str(uuid.uuid4()),
+                                           query,
+                                           query_parameters=query_parameters)
+        query_job = _set_common_query_settings(query_job)
+        query_job.begin()
+        try:
+            query_results = query_job.results().fetch_data()
+        except:
+            query_results = query_job.result().fetch_data()
+        if verbose:
+            print("Query running...")
+        _wait_for_job(query_job)
+        if verbose:
+            print("Query done.")
+            if query_job._properties["statistics"]["query"].get("cacheHit",
+                                                                False):
+                print("Cache hit.")
+            elif ("statistics" in query_job._properties and
+                    "query" in query_job._properties["statistics"]):
+                bytes_billed = int(query_job
+                                   ._properties["statistics"]["query"]
+                                   .get("totalBytesProcessed", 0))
+                bytes_processed = int(query_job
+                                      ._properties["statistics"]["query"]
+                                      .get("totalBytesBilled", 0))
+                print("Total bytes billed (processed): %s (%s)" %
+                      (sizeof_fmt(bytes_billed), sizeof_fmt(bytes_processed)))
+            print("\nRetrieving results...")
+        return query_results, query_job
+
+    if (configuration and "timeout_ms" in configuration) or timeout_ms:
+        query_results, query_job = sync_query()
+        rows = list(query_results.rows)
+        total_rows = len(rows)
     else:
-        final_df = _parse_data(schema, [])
+        query_results, query_job = async_query()
+        rows = list(query_results)
+        total_rows = len(rows)
+
+    if verbose:
+        print("Got %s rows.") % total_rows
+        if query_job:
+            print("\nTotal time taken %ss" % (datetime.utcnow() -
+                  query_job.created.replace(tzinfo=None)).seconds)
+            print("Finished at %s." % datetime.now()
+                  .strftime('%Y-%m-%d %H:%M:%S'))
+
+    if return_type == 'schema':
+        return query_results.schema
+    elif return_type == 'list':
+        return rows
+
+    columns = [field.name for field in query_results.schema]
+    data = rows
+
+    final_df = DataFrame(data=data, columns=columns)
+
```

Comment:
it seems that you are defining all of these functions *within* the read_gbq function? why is that

Reply:
I'm replicating how it was written in the previous version of read_gbq, which similarly handled re-ordering and type conversion all within read_gbq. If you think this should be refactored, I can take a crack at it, but perhaps it can be left for another day? The schema/column names, I can try splitting out into another function to roughly mirror how it was done before if that makes sense.